### ISSUE TITLE
[BUG] some getByRole roles don't work

### ISSUE THREAD

**Issue by eatonjl** [microsoft/playwright#23377]
### System info
- Playwright Version: 1.33
- Operating System: Mac OS
- Browser: Chromium
- Other info:

### Source code

- [x] I provided exact source code that allows reproducing the issue locally.

**Config file**

```ts
import { defineConfig, devices } from '@playwright/test';

/**
 * Read environment variables from file.
 * https://github.com/motdotla/dotenv
 */
// require('dotenv').config();

/**
 * See https://playwright.dev/docs/test-configuration.
 */
export default defineConfig({
  testDir: './tests',
  /* Run tests in files in parallel */
  fullyParallel: true,
  /* Fail the build on CI if you accidentally left test.only in the source code. */
  forbidOnly: !!process.env.CI,
  /* Retry on CI only */
  retries: process.env.CI ? 2 : 0,
  /* Opt out of parallel tests on CI. */
  workers: process.env.CI ? 1 : undefined,
  /* Reporter to use. See https://playwright.dev/docs/test-reporters */
  reporter: 'html',
  /* Shared settings for all the projects below. See https://playwright.dev/docs/api/class-testoptions. */
  use: {
    /* Base URL to use in actions like `await page.goto('/')`. */
    // baseURL: 'http://127.0.0.1:3000',

    /* Collect trace when retrying the failed test. See https://playwright.dev/docs/trace-viewer */
    trace: 'on',
  },

  /* Configure projects for major browsers */
  projects: [
    {
      name: 'chromium',
      use: { ...devices['Desktop Chrome'] },
    },

    {
      name: 'firefox',
      use: { ...devices['Desktop Firefox'] },
    },

    {
      name: 'webkit',
      use: { ...devices['Desktop Safari'] },
    },

    /* Test against mobile viewports. */
    // {
    //   name: 'Mobile Chrome',
    //   use: { ...devices['Pixel 5'] },
    // },
    // {
    //   name: 'Mobile Safari',
    //   use: { ...devices['iPhone 12'] },
    // },

    /* Test against branded browsers. */
    // {
    //   name: 'Microsoft Edge',
    //   use: { ...devices['Desktop Edge'], channel: 'msedge' },
    // },
    // {
    //   name: 'Google Chrome',
    //   use: { ..devices['Desktop Chrome'], channel: 'chrome' },
    // },
  ],

  /* Run your local dev server before starting the tests */
  // webServer: {
  //   command: 'npm run start',
  //   url: 'http://127.0.0.1:3000',
  //   reuseExistingServer: !process.env.CI,
  // },
});
```

**Test file (self-contained)**

```ts
import { test, expect } from '@playwright/test';

test('getByRoleIssue', async ({ page }) => {
  await page.goto('https://www.google.com/search?q=pizza+hut&oq=pizza+hut&aqs=chrome..69i57.1522j0j9&sourceid=chrome&ie=UTF-8');
  await page.getByRole('link', { name: 'Order Online', exact: true }).click();
  await page.getByRole('caption', { name: 'See More'}).click();
});
```

**Steps**
- Place a debug step on the line "await page.getByRole('caption', { name: 'See More'}).click();"
- When the test hits the debug line, go to the page, inspect it, and add "role=caption" to the "See More" link on the page
- Continue the test

Next
Do the same thing but add "role=button" instead

**Expected**

both tests should pass

**Actual**

The first one (with "caption") fails while the second one (with "button") passes. For some reason, captions don't work even though they are listed as an option in the getbyrole description. I have gotten caption to work when I only look by role. But if I look by role AND name, it fails.

**Comment by eatonjl** [comment]
On line 2615 of the types.d.ts file it says, "getByRole(role: "`alert"|"alertdialog"|"application"|"article"|"banner"|"blockquote"|"button"|"caption"|"cell"|"checkbox"|"code"|"columnheader"|"combobox"|"complementary"|"contentinfo"|"definition"|"deletion"|"dialog"|"directory"|"document"|"emphasis"|"feed"|"figure"|"form"|"generic"|"grid"|"gridcell"|"group"|"heading"|"img"|"insertion"|"link"|"list"|"listbox"|"listitem"|"log"|"main"|"marquee"|"math"|"meter"|"menu"|"menubar"|"menuitem"|"menuitemcheckbox"|"menuitemradio"|"navigation"|"none"|"note"|"option"|"paragraph"|"presentation"|"progressbar"|"radio"|"radiogroup"|"region"|"row"|"rowgroup"|"rowheader"|"scrollbar"|"search"|"searchbox"|"separator"|"slider"|"spinbutton"|"status"|"strong"|"subscript"|"superscript"|"switch"|"tab"|"table"|"tablist"|"tabpanel"|"term"|"textbox"|"time"|"timer"|"toolbar"|"tooltip"|"tree"|"treegrid"|"treeitem", options?: {`"
You can see that "caption" is on the list.

**Comment by thomas-phillips-nz** [comment]
From my testing, the `name` property doesn't match when the role is set to `"caption"`.

Setting the locator to `await page.getByRole('caption').click();` seems to work fine, so Playwright can find the element okay.

That said, what are you trying to achieve? You're explicitly adding `role="caption"` to the DOM, are you having issues with other caption locators? How does Playwright recognise them when using the "Pick locator" option in vscode?

**Comment by eatonjl** [comment]
Thanks for looking into it! The reason I am explicitly adding role=caption is purely for re-creation purposes. The code I am working on is not easily sharable and I didn't want to spend hours trying to find a site that had an element with role=caption. I figured this was the easiest way to reproduce the issue. I have code in our application that has role=caption and this is the outcome when testing. And you are correct about `await page.getByRole('caption').click();` working. As I stated in my initial post, navigating to just the role works but trying to use a name with that specific role does not whereas it does with role=button and others. There is no mention in the documentation that some roles will work with names and others won't so I assume there could be a bug. When using the "pick locator" option it will recommend `page.getbytext()`, indicating it does not recognize the role. If I change the role to a button then it recommends `page.getbyrole(button, {name: [name]};`.
I did notice that Edge has an auto-complete in the inspector's element tab that shows you a list of role options. That list has the following: `alert; application; article; banner; button; cell; checkbox; complementary; contentinfo; dialog; document; feed; figure; form; grid; gridcell; heading; img; list; listbox; listitem; main; navigation; region; row; rowgroup; search; switch; tab; table; tabpanel; textbox; timer`. 
This means that the following may not be supported anymore? 
`alertdialog; blockquote; caption; code; columnheader; combobox; definition; deletion; directory; emphasis; generic; group; insertion; link; log; marquee; math; meter; menu; menubar; menuitem; menuitemcheckbox; menuitemradio; none; note; option; paragraph; presentation; progressbar; radio; radiogroup; rowheader; scrollbar; searchbox; separator; slider; spinbutton; status; strong; subscript; tablist;  superscript; term; time; toolbar; tooltip; tree; treegrid; treeitem`

**Comment by dgozman** [comment]
Investigation notes. For the following snippet, there are various opinions for the roles/names.

```html
<a id="link1" href="example.com" role="caption">NAME1</a>
<a id="link2" href="example.com" role="caption" aria-label="LABEL2">NAME2</a>
<a id="link3" href="example.com" role="term">NAME3</a>
<a id="link4" href="example.com" role="term" aria-label="LABEL4">NAME4</a>
<a id="link5" href="example.com" role="presentation">NAME5</a>
<a id="link6" href="example.com" role="presentation" aria-label="LABEL6">NAME6</a>
```

- Chromium: "NAME1", "LABEL2", "", "LABEL4", "NAME5" (role=link), "LABEL6" (role=link)
- Firefox: "", "LABEL2", "NAME3", "LABEL4", "NAME5" (role=link), "LABEL6" (role=link)
- WebKit: "", "LABEL2", "", "LABEL4", "NAME5" (role=link), "LABEL6" (role=link)
- ARIA spec: "", "LABEL2", "", "LABEL4", "" (role=presentation), "LABEL6" (role=link)

**Comment by gauthierm** [comment]
I have this problem using the `searchbox` role. If I inspect the element in Firefox, it shows the role is `textbox` and the name is `Search My Site`. If I inspect in Chrome, it shows `searchbox` and `Search My Site`.

In Playwright, if I use `await expect(page.getByRole('searchbox')).toBeAttached();` it works. If I use `await expect(page.getByRole('searchbox', { name: 'Search My Site' })).toBeAttached();` it fails.

I expect it will also fail in Firefox as the detected role is different in the a11y inspector.

Using `getByRole('link', { name: 'Test' })` works great.

**Comment by kristojorg** [comment]
Role `alert` also seems affected by this. I can select an element with `getByRole("alert")` but am unable to select the element with `getByRole("alert", { name: "<string child contents>" })`

**Comment by MaruschkaScheepersW** [comment]
This issue seems to affect combobox as well. 

I can select the combobox just fine with 
`this.portalSwitch = page.getByRole('combobox').first();`

But as soon as I specify the name, it fails to select the element.

**Comment by dgozman** [comment]
@kristojorg @MaruschkaScheepersW Could you please file a new issue by filling in the "Bug Report" template and providing a repro? ARIA is full of tricky details, so we need the exact scenario to make sure there is a proper fix.

**Comment by jdgarvey** [comment]
@dgozman I also ran into this issue with the combobox, I submitted a new issue with a minimal reproduction here: https://github.com/microsoft/playwright/issues/31135.

**Comment by luchsamapparat** [comment]
I stumbled upon this link in another thread regarding the `definition` role. It seems, that both `definition` and `caption` both cannot be named, even though the Chrome DevTools says otherwise.

https://w3c.github.io/aria/#namefromprohibited

**Comment by pavelfeldman** [comment]
https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/structural_roles suggests that this role can't be named, consistent with ^^. Closing.

**Comment by npup** [comment]
Reading in the MDN pages about the role term, for example (https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Reference/Roles/term_role#description), it sure says

> Warning: The accessible name should be the term itself, so DO NOT use aria-label or aria-labelledby.

I have this:

`<span role="term">Foo</span>`

and still cannot target the above element with `page.getByRole("term", { name: "Foo" })` - which I think _should_ work. When I tested using (React) Testing Library, such selectors worked fine.

**Comment by jasperdunn** [comment]
This is what I used as a workaround:

```ts
await expect(detailsCard.locator('dt:has-text("Year") + dd')).toHaveText(
    `${updatedYear}`,
  );
```

**Comment by GC-BenLeeHughes** [comment]
When MDN is saying one thing and the WAI is saying another, and most systems recognize names on terms and such other roles, the practicality of Playwright not finding terms by name is frustrating. Perhaps maybe not by name then, but maybe we could for roles by direct text instead?

**Comment by GC-BenLeeHughes** [comment]
I'm having a professional conniption over here.
https://issues.chromium.org/issues/381418589

Chromium explicitly added names to the terms even though the request was for names to show correctly for the _parent_ element when a term was inside it. This very likely is why Chromium is showing names for terms.

**Comment by GC-BenLeeHughes** [comment]
@jasperdunn and anyone else here, I have a Chromium issue regarding names existing on "term" roles filed here:
https://issues.chromium.org/issues/434005975

This is a regression in Chome caused by issue by this issue being fixed by incorrectly breaking WAI spec:
https://issues.chromium.org/issues/381418589

**Comment by jasperdunn** [comment]
> @jasperdunn and anyone else here, I have a Chromium issue regarding names existing on "term" roles filed here:
> https://issues.chromium.org/issues/434005975
> 
> This is a regression in Chome caused by issue by this issue being fixed by incorrectly breaking WAI spec:
> https://issues.chromium.org/issues/381418589

Oh hell yeah, chromium issues actually get worked on! TIL 🙏😅

**Comment by GC-BenLeeHughes** [comment]
Aye. They do appear to investigate regressions. This does mean that terms and definitions don't recieve names, though, as per spec, so the answer you gave regarding the locator and sibling element appears to be the best solution here. I've looked around for other solutions to this, but this seems to be it.
